server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4484,9 +4484,30 @@ public Durability getTranslogDurability() {
     // we can not protect with a lock since we "release" on a different thread
     private final AtomicBoolean flushOrRollRunning = new AtomicBoolean();
 
+    /**
+     * Checks if the shard need to be Refreshed depending on Translog constraints.
+     * Each Translog type can have it's own decider
+     * @return {@code true} if the shard should be Refreshed
+     */
+    public boolean shouldRefreshShard() {
+        final Engine engine = getEngineOrNull();
+        if (engine != null) {
+            try {
+                return engine.translogManager().shouldRefreshShard();
+            } catch (final AlreadyClosedException e) {
+                // we are already closed, no need to Refresh
+            }
+        }
+        return false;
+    }
+
+    private final AtomicBoolean isRefreshRunning = new AtomicBoolean();
```

Comment:
why is this check needed here, what happens today to prevent concurrent refresh executions?

Reply:
Here we are trying to call a blocking refresh, which will definitely trigger the Refresh (by waiting for already running refresh to complete). Since we don't want to trigger multiple refreshes in the background, because a Refresh will take some time to complete and Trim-translog will also take some time. In case there are writes happening so fast we might end up scheduling so many Refreshes (each using a different Refresh Thread, consuming our resources), before the first one completes and trimming gets done.

Reply:
Removed this, and going ahead with non-blocking refreshes. Had discussion with @gbbafna.